Modules/_pickle.c

```diff
@@ -1601,15 +1601,15 @@ memo_get(PicklerObject *self, PyObject *key)
 /* Store an object in the memo, assign it a new unique ID based on the number
    of objects currently stored in the memo and generate a PUT opcode. */
 static int
-memo_put(PicklerObject *self, PyObject *obj)
+memo_put(PicklerObject *self, PyObject *obj, int opt)
```

Comment:
I don't understand how you decide whether `opt` should be 0 or 1. What is the heuristic?

Reply:
+1 on this. This is critical behavior for `cloudpickle` :)

Reply:
This is an interesting question. The rule is opt=0 for objects which save its content after saving itself. They are non-empty lists, sets, dicts and general objects with non-trivial elements 2-4 of the tuple returned by `__reduce__()`. They should be memoized to allow detecting reference loops.